ffi/src/lib.rs

```diff
@@ -1175,43 +1226,93 @@ impl Drop for ClusterScanCursor {
 /// * `channel` must be valid until it is passed in a call to [`free_command_response`].
 /// * Both the `success_callback` and `failure_callback` function pointers need to live while the client is open/active. The caller is responsible for freeing both callbacks.
 #[no_mangle]
-pub unsafe extern "C" fn request_cluster_scan(
+pub unsafe extern "C-unwind" fn request_cluster_scan(
     client_adapter_ptr: *const c_void,
     channel: usize,
     cursor: ClusterScanCursor,
-    arg_count: c_ulong,
-    args: *const usize,
-    args_len: *const c_ulong,
+    arg_count: usize,
+    args: *const *const u8,
+    args_len: *const usize,
 ) -> *mut CommandResult {
-    let client_adapter =
-        unsafe { Box::leak(Box::from_raw(client_adapter_ptr as *mut ClientAdapter)) };
+    let client_adapter = unsafe {
+        // we increment the strong count to ensure that the client is not dropped just because we turned it into an Arc.
+        Arc::increment_strong_count(client_adapter_ptr);
+        Arc::from_raw(client_adapter_ptr as *mut ClientAdapter)
+    };
     let c_str = unsafe { CStr::from_ptr(cursor.get_cursor()) };
-    let temp_str = c_str.to_str().expect("Must be UTF-8");
+    let temp_str = match c_str.to_str() {
+        Ok(temp_str) => temp_str,
+        Err(e) => return client_adapter.handle_error(RedisError::from(e), channel),
+    };
     let cursor_id = temp_str.to_string();
 
     let cluster_scan_args: ClusterScanArgs = if arg_count > 0 {
-        let arg_vec = unsafe {
-            convert_double_pointer_to_vec(args as *const *const c_void, arg_count, args_len)
-        };
+        let arg_vec = unsafe { convert_double_pointer_to_vec(args, arg_count, args_len) };
 
         let mut pattern: &[u8] = &[];
         let mut object_type: &[u8] = &[];
         let mut count: &[u8] = &[];
 
-        for i in 0..arg_count as usize {
-            match arg_vec[i] {
-                b"MATCH" => pattern = arg_vec[i + 1],
-                b"TYPE" => object_type = arg_vec[i + 1],
-                b"COUNT" => count = arg_vec[i + 1],
-                _ => {}
+        debug_assert!(arg_count <= arg_vec.len(), "arg_count greater than arg_vec length. This is probably due to a bug in convert_double_pointer_to_vec.");
+
+        if arg_count < usize::MAX {
```

Comment:
I think this command has up to 6 args, where count is even. You can do such check on line 1249 and remove this `if`.

Reply:
I don't think this `if` can be avoided. I can change the count check, but `request_cluster_scan` is still a public extern function and can be called with invalid arguments by anyone.